### PR TITLE
[Autotune] Use half the number of warps for  reduction tuning on AMD.

### PR DESCRIPTION
**Pull request by htyu**
I was seeing for a reduction kernel and a given block size, on AMDGPU, the vectorization bandwidth (16-byte) for a thread was not fully leveraged while it was not a problem for NVGPU. It appeared that each thread got fewer data to process as a whole row were processed by more threads, and the number of elements each thread got was not enough to saturate full vectorization. On AMDGPU, a warp has 64 lanes compared to 32 on the NV side. Therefore I'm tuning down the default number of warps (8 for NV) for AMD. I'm seeing 10% speed up for an internal benchmark.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

**Comment by htyu** [comment]
@shunting314 please take a look to see if the change makes sense. There are other places using 8 as num_warps. I haven't got a chance to investigate the codegne there, so wondering if they should be changed as well. Thanks.

**Comment by shunting314** [comment]
Can you share the example kernel that motivates this PR?

yea, similar change of triton config heuristics probably should be done for pointwise and persistent-reduction as well for AMDGPU. It would be great if we can have examples to show the perf difference for piontwise/persisrent-reduction based on your examples for non-persistent reduction.

**Review by shunting314 — APPROVED** [review]

**Comment by htyu** [comment]
> Can you share the example kernel that motivates this PR?
> 

Sure, here is the kernel: https://gist.github.com/htyu/91d929259f8a1f9958c160c6ad8cb184


TTGIR before: https://gist.github.com/htyu/00e88346f18e687570180c74cf356276

TTGIR after: https://gist.github.com/htyu/f47c4f00d6ab04bb34fc86f330d45585

And you can see for same amount of data, in old ttgir there is `sizePerThread = [2]` which translates into a 32-bit load, while `sizePerThread = [4]` in new ttgir that translates into a 64-bit load.

> yea, similar change of triton config heuristics probably should be done for pointwise and persistent-reduction as well for AMDGPU. It would be great if we can have examples to show the perf difference for piontwise/persisrent-reduction based on your examples for non-persistent reduction.

I don't have for now but I'll keep an eye on it.

File: torch/_inductor/runtime/triton_heuristics.py [hunk 1]
```diff
@@ -1272,7 +1272,10 @@ def triton_config_reduction(size_hints, x, r, num_stages=1, num_warps=None) -> C
     cfg = {"XBLOCK": x, "RBLOCK": r}
     if num_warps is None:
         num_warps = conditional_product(x, r) // 128
-    num_warps = next_power_of_2(min(max(num_warps, 2), 8))
+    # On AMD GPU each warp has 64 lanes which is double the size on NV GPU,
+    # therefore using half the number of warps here correspondingly.
+    default_num_warps = 4 if torch.version.hip else 8
+    num_warps = next_power_of_2(min(max(num_warps, 2), default_num_warps))
```
[the remark this answers is not in the document]
Comment (htyu) [reply torch/_inductor/runtime/triton_heuristics.py:1278]: Yeah, makes sense to change it to 1, though it doesn't affect my case.

**Comment by shunting314** [comment]
> And you can see for same amount of data, in old ttgir there is sizePerThread = [2] which translates into a 32-bit load, while sizePerThread = [4] in new ttgir that translates into a 64-bit load.

I see. This is better than before but this is still not a full vectorization since each load handles 8 bytes rather than 16 bytes. I'm wondering how the perf can look like if sizePerThread = [8] so we do full vectorization here. (note dtype is bf16 here).

**Comment by htyu** [comment]
> > And you can see for same amount of data, in old ttgir there is sizePerThread = [2] which translates into a 32-bit load, while sizePerThread = [4] in new ttgir that translates into a 64-bit load.
> 
> I see. This is better than before but this is still not a full vectorization since each load handles 8 bytes rather than 16 bytes. I'm wondering how the perf can look like if sizePerThread = [8] so we do full vectorization here. (note dtype is bf16 here).

Good point. I believe it's going to be even better. I was thinking about that too. This is also a problem on the NV side, i.e, given a size hint, `size_hints=[512, 1024],` in this case, the compiler didn't estimate accurately how much data each thread should get. It sounds like a more general issue and I'll take a look, perhaps separate to this patch.

**Pull request closed without merging**